packages/oauth/oauth-provider/src/dpop/dpop-manager.ts

```diff
@@ -47,110 +48,153 @@ export class DpopManager {
-      typeof payload['nonce'] !== 'string'
-    ) {
-      throw new InvalidDpopProofError('DPoP nonce must be a string')
+    // Note rfc9110#section-9.1 states that the method name is case-sensitive
+    if (!htm || htm !== httpMethod) {
+      throw new InvalidDpopProofError('DPoP "htm" mismatch')
     }
 
-    if (!payload['nonce'] && this.dpopNonce) {
-      throw new UseDpopNonceError()
+    if (!htu || typeof htu !== 'string') {
+      throw new InvalidDpopProofError('Invalid DPoP "htu" type')
     }
 
-    if (payload['nonce'] && !this.dpopNonce?.check(payload['nonce'])) {
-      throw new UseDpopNonceError('DPoP nonce mismatch')
+    // > To reduce the likelihood of false negatives, servers SHOULD employ
+    // > syntax-based normalization (Section 6.2.2 of [RFC3986]) and
+    // > scheme-based normalization (Section 6.2.3 of [RFC3986]) before
+    // > comparing the htu claim.
+    //
+    // RFC9449 section 4.3. Checking DPoP Proofs - https://datatracker.ietf.org/doc/html/rfc9449#section-4.3
+    if (!htu || normalizeHtu(htu) !== urlToHtu(httpUrl)) {
+      throw new InvalidDpopProofError('DPoP "htu" mismatch')
     }
 
-    const htuNorm = normalizeHtu(htu)
-    if (!htuNorm) {
-      throw new TypeError('Invalid "htu" argument')
+    if (!nonce && this.dpopNonce) {
+      throw new UseDpopNonceError()
     }
 
-    if (htuNorm !== normalizeHtu(payload['htu'])) {
-      throw new InvalidDpopProofError('DPoP htu mismatch')
+    if (nonce && !this.dpopNonce?.check(nonce)) {
+      throw new UseDpopNonceError('DPoP "nonce" mismatch')
     }
 
     if (accessToken) {
-      const athBuffer = createHash('sha256').update(accessToken).digest()
-      if (payload['ath'] !== athBuffer.toString('base64url')) {
-        throw new InvalidDpopProofError('DPoP ath mismatch')
+      const accessTokenHash = createHash('sha256').update(accessToken).digest()
+      if (ath !== accessTokenHash.toString('base64url')) {
+        throw new InvalidDpopProofError('DPoP "ath" mismatch')
       }
-    } else if (payload['ath']) {
-      throw new InvalidDpopProofError('DPoP ath not allowed')
+    } else if (ath !== undefined) {
+      throw new InvalidDpopProofError('DPoP "ath" claim not allowed')
     }
 
-    try {
-      return {
-        protectedHeader,
-        payload,
-        jkt: await calculateJwkThumbprint(protectedHeader['jwk']!, 'sha256'), // EmbeddedJWK
-      }
-    } catch (err) {
-      const message =
-        err instanceof JOSEError ? err.message : 'Failed to calculate jkt'
-      throw new InvalidDpopProofError(message, err)
-    }
+    // @NOTE we can assert there is a jwk because the jwtVerify used the
+    // EmbeddedJWK key getter mechanism.
+    const jwk = protectedHeader.jwk!
+    const jkt = await calculateJwkThumbprint(jwk, 'sha256').catch((err) => {
+      throw err instanceof JOSEError
+        ? new InvalidDpopProofError(err.message)
+        : new InvalidDpopProofError('Failed to calculate jkt', err)
```

Comment:
There's a lot going on here, I'd recommend a more explicit construction.
```suggestion
      if (err instanceof JOSEError) {
        throw new InvalidDpopProofError(err.message)
      } else {
        throw new InvalidDpopProofError('Failed to calculate jkt', err)
      }
```